File: app/src/main/java/org/sil/storyproducer/media/pipe/PipedAudioMixer.java
package org.sil.storyproducer.media.pipe;

import android.media.MediaCodec;
import android.media.MediaFormat;

import org.sil.storyproducer.media.MediaHelper;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.ShortBuffer;
import java.util.ArrayList;
import java.util.List;

/**
 * <p>This media pipeline component mixes raw audio streams together.</p>
 * <p>This component also optionally changes the volume of the raw audio stream.</p>
 */
public class PipedAudioMixer extends PipedAudioShortManipulator implements PipedMediaByteBufferDest {
    private static final String TAG = "PipedAudioMixer";

    private MediaFormat mOutputFormat;

    private boolean mIsDone = false;

    private List<PipedMediaByteBufferSource> mSources = new ArrayList<>();
    private List<Float> mSourceVolumeModifiers = new ArrayList<>();
    private List<ByteBuffer> mSourceBuffers = new ArrayList<>();
    private List<ShortBuffer> mSourceShortBuffers = new ArrayList<>();

    private MediaCodec.BufferInfo mInfo = new MediaCodec.BufferInfo();

    public PipedAudioMixer() {
        //empty default constructor
    }

    @Override
    public void addSource(PipedMediaByteBufferSource src) throws SourceUnacceptableException {
        addSource(src, 1);
    }

    /**
     * Specify a predecessor of this component in the pipeline with a specified volume scaling factor.
     * @param src the preceding component of the pipeline.
     * @param volumeModifier volume scaling factor.
     * @throws SourceUnacceptableException if source is null.
     */
    public void addSource(PipedMediaByteBufferSource src, float volumeModifier) throws SourceUnacceptableException {
        if(src == null) {
            throw new SourceUnacceptableException("Source cannot be null!");
        }

        mSources.add(src);
        mSourceVolumeModifiers.add(volumeModifier);
        mSourceBuffers.add(null);
        mSourceShortBuffers.add(null);
    }

    @Override
    public void setup() throws IOException, SourceUnacceptableException {
        for (int i = 0; i < mSources.size(); i++) {
            PipedMediaByteBufferSource source = mSources.get(i);
            source.setup();
            MediaFormat format = source.getOutputFormat();

            if (source.getMediaType() != MediaHelper.MediaType.AUDIO) {
                throw new SourceUnacceptableException("Source must be audio!");
            }

            if(!format.getString(MediaFormat.KEY_MIME).equals(MediaHelper.MIMETYPE_RAW_AUDIO)) {
                throw new SourceUnacceptableException("Source audio must be a raw audio stream!");
            }

            if (mChannelCount == 0) {
                mChannelCount = format.getInteger(MediaFormat.KEY_CHANNEL_COUNT);
            } else if (mChannelCount != format.getInteger(MediaFormat.KEY_CHANNEL_COUNT)) {
                throw new SourceUnacceptableException("Source audio channel counts don't match!");
            }

            if (mSampleRate == 0) {
                mSampleRate = format.getInteger(MediaFormat.KEY_SAMPLE_RATE);
            } else if (mSampleRate != format.getInteger(MediaFormat.KEY_SAMPLE_RATE)) {
                throw new SourceUnacceptableException("Source audio sample rates don't match!");
            }

            fetchSourceBuffer(i);
        }

        mOutputFormat = MediaHelper.createFormat(MediaHelper.MIMETYPE_RAW_AUDIO);
        mOutputFormat.setInteger(MediaFormat.KEY_SAMPLE_RATE, mSampleRate);
        mOutputFormat.setInteger(MediaFormat.KEY_CHANNEL_COUNT, mChannelCount);
    }

    @Override
    public MediaFormat getOutputFormat() {
        return mOutputFormat;
    }

    @Override
    public boolean isDone() {
        return mIsDone;
    }

    protected short getSampleForTime(long time, int channel) {
        int sum = 0;

        //Loop through all sources and add samples.
        for(int i = 0; i < mSources.size(); i++) {
            ShortBuffer sBuffer = mSourceShortBuffers.get(i);
            float volumeModifier = mSourceVolumeModifiers.get(i);
            while(sBuffer != null && sBuffer.remaining() <= 0) {
                releaseSourceBuffer(i);
                fetchSourceBuffer(i);
                sBuffer = mSourceShortBuffers.get(i);
            }
            if(sBuffer != null) {
                sum += sBuffer.get() * volumeModifier;
            }
            else {
                //Remove depleted sources from the lists.
                mSources.remove(i);
                mSourceBuffers.remove(i);
                mSourceShortBuffers.remove(i);

                //Decrement i so that former source i + 1 is not skipped.
                i--;
            }
        }

        //If sources are all gone, this component is done.
        if(mSources.isEmpty()) {
            mIsDone = true;
        }

        //Clip the sum to a short.
        return (short) sum;
    }

    private void fetchSourceBuffer(int sourceIndex) {
        PipedMediaByteBufferSource source = mSources.get(sourceIndex);
        if(source.isDone()) {
            return;
        }
        ByteBuffer buffer = source.getBuffer(mInfo);
        mSourceBuffers.set(sourceIndex, buffer);
        mSourceShortBuffers.set(sourceIndex, MediaHelper.getShortBuffer(buffer));
    }

    private void releaseSourceBuffer(int sourceIndex) {
        mSources.get(sourceIndex).releaseBuffer(mSourceBuffers.get(sourceIndex));
        mSourceBuffers.set(sourceIndex, null);
        mSourceShortBuffers.set(sourceIndex, null);
    }

    @Override
    public void close() {
        super.close();
        for(PipedMediaByteBufferSource source : mSources) {
            source.close();
        }
    }
}

File: app/src/main/java/org/tyndalebt/spadv/tools/media/pipe/PipedAudioMixer.kt
package org.tyndalebt.spadv.tools.media.pipe

import android.media.MediaFormat
import android.util.Log
import org.tyndalebt.spadv.tools.media.MediaHelper
import java.io.IOException
import java.util.*
import kotlin.math.min

/**
 *
 * This media pipeline component mixes raw audio streams together.
 *
 * This component also optionally changes the volume of the raw audio stream.
 */
class PipedAudioMixer : PipedAudioShortManipulator(), PipedMediaByteBufferDest {
    override val componentName: String
        get() = TAG

    private var mOutputFormat: MediaFormat? = null

    private val mixSources = ArrayList<PipedMediaByteBufferSource>()
    private val mixBuffers = ArrayList<ShortArray>()
    private val mixPoss = ArrayList<Int>()
    private val mixEnds = ArrayList<Int>()

    private var mCurrentSample: ShortArray? = null

    override fun getOutputFormat(): MediaFormat? {
        return mOutputFormat
    }

    /**
     * Specify a predecessor of this component in the pipeline with a specified volume scaling factor.
     * @param src the preceding component of the pipeline.
     * @throws SourceUnacceptableException if source is null.
     */
    @Throws(SourceUnacceptableException::class)
    override fun addSource(src: PipedMediaByteBufferSource?) {
        if (src == null) {
            throw SourceUnacceptableException("Source cannot be null!")
        }

        mixSources.add(src)
    }

    @Throws(IOException::class, SourceUnacceptableException::class)
    override fun setup() {
        if (mComponentState != PipedMediaSource.State.UNINITIALIZED) {
            return
        }

        if (mixSources.isEmpty()) {
            throw SourceUnacceptableException("No sources specified!")
        }

        for (i in mixSources.indices) {
            val source = mixSources[i]
            source.setup()
            validateSource(source, mChannelCount, mSampleRate)

            val format = source.outputFormat
            if (mChannelCount == 0) {
                mChannelCount = format.getInteger(MediaFormat.KEY_CHANNEL_COUNT)
            }
            if (mSampleRate == 0) {
                mSampleRate = format.getInteger(MediaFormat.KEY_SAMPLE_RATE)
            }

            mixBuffers.add(ShortArray(MediaHelper.MAX_INPUT_BUFFER_SIZE / 2))
            mixPoss.add(0)
            mixEnds.add(0)
            try {
                fetchSourceBuffer(i)
            } catch (e: SourceClosedException) {
                //This case should not happen.
                throw SourceUnacceptableException("First fetchSourceBuffer failed! Strange", e)
            }

        }

        mOutputFormat = MediaHelper.createFormat(MediaHelper.MIMETYPE_RAW_AUDIO)
        mOutputFormat!!.setInteger(MediaFormat.KEY_SAMPLE_RATE, mSampleRate)
        mOutputFormat!!.setInteger(MediaFormat.KEY_CHANNEL_COUNT, mChannelCount)

        mCurrentSample = ShortArray(mChannelCount)

        mComponentState = PipedMediaSource.State.SETUP

        start()
    }

    @Throws(SourceClosedException::class)
    override fun loadSamples(): Boolean {
        for (i in 0 until mChannelCount) {
            mCurrentSample!![i] = 0
        }

        //Loop through all sources and fetch buffers if we need to
        var iSource = 0
        var allLength = 1000000
        while (iSource < mixSources.size) {
            if (mixPoss[iSource] >= mixEnds[iSource] && mixBuffers.getOrNull(iSource) != null) {
                fetchSourceBuffer(iSource)
            }
            //The source is depleted!
            if (mixBuffers.getOrNull(iSource) == null) {
                //Remove depleted sources from the lists.
                mixSources.removeAt(iSource)
                mixPoss.removeAt(iSource)
                mixEnds.removeAt(iSource)
                continue
            }
            allLength = min(allLength,mixEnds[iSource] - mixPoss[iSource])
            iSource++
        }

        if(mixSources.isEmpty()) return false

        //prep srcBuffer with the mixing of all channels available.

        //do the first channel
        //grab the buffer

        //setup the data as a source.  Copy the buffer

        srcBuffer = mixBuffers[0].copyOfRange(mixPoss[0],mixPoss[0]+allLength)
        mixPoss[0] += allLength

        //do all other channels
        iSource = 1
        while (iSource < mixSources.size) {
            //grab the buffer
            val pos = mixPoss[iSource]
            val mb = mixBuffers[iSource].sliceArray(pos..pos+allLength)

            //setup the data as a source.  Copy the buffer
            srcBuffer.forEachIndexed { index, sh ->  srcBuffer[index] = (sh + mb[index]).toShort()}

            mixPoss[iSource] += allLength
            iSource++
        }

        srcPos = 0
        srcEnd = allLength

        return true
    }

    @Throws(SourceClosedException::class)
    fun fetchSourceBuffer(sourceIndex: Int) {
        val source = mixSources[sourceIndex]
        if (source.isDone) {
            source.close()
            mixBuffers.removeAt(sourceIndex)
            return
        }

        //buffer of bytes
        val buffer = source.getBuffer(mInfo)
        //buffer of shorts (16-bit samples)
        val sBuffer = MediaHelper.getShortBuffer(buffer)

        if (MediaHelper.VERBOSE) {
            Log.v(TAG, "Received " + (if (buffer.isDirect) "direct" else "non-direct")
                    + " buffer of size " + mInfo.size
                    + " with" + (if (buffer.hasArray()) "" else "out") + " array")
        }

        val pos = 0
        val size = sBuffer.remaining()
        //Copy ShortBuffer to array of shorts in hopes of speedup.
        sBuffer.get(mixBuffers[sourceIndex], pos, size)
        mixPoss[sourceIndex] = pos
        mixEnds[sourceIndex] = size

        //Release buffer since data was copied.
        source.releaseBuffer(buffer)
    }

    override fun close() {
        super.close()
        while (!mixSources.isEmpty()) {
            val source = mixSources.removeAt(0)
            source.close()
        }
    }

    companion object {
        private val TAG = "PipedAudioMixer"
    }
}